Adds basic docs for UI authentication [OEYE-59] (#44)
banzaicloud/one-eye-docs: docs/authentication/_index.md

---
title: Authentication on the UI
shorttitle: Authentication
weight: 800
---

You can secure One Eye access using an external ingress, and authenticate and authorize user connections. The solution is based on [Pomerium](https://github.com/pomerium/pomerium) and [Dex](https://github.com/dexidp/dex), allowing you to use the identity provider of your choice (for example, GitHub or LDAP). You can also apply role based authorization, and assign read-only or read and write access to One Eye.

> Authentication uses cert-manager to establish secure connections. Authentication is performed using Pomerium, which uses Dex to authenticate requests to external OIDC providers.

Reviewer: ```suggestion
Authentication uses the latest stable version of cert-manager to ensure secure TLS certificates for external components like Pomerium and Dex. Authentication is performed by Pomerium with the help of Dex through OIDC to authenticate requests to external identity providers.

Pomerium applies route based access policies so that general access to specific components can be gated at the proxy level.

Once the request successfully passes Pomerium, the One Eye Dashboard will check group-role mapping information to identify which groups should have viewer and/or editor access on the UI. 

> Note: This doesn't apply to provided components, like Prometheus or the Grafana dashboard, where all users will be viewers. In case you need Grafana administrator access, use the `one-eye ingress connect` command to access the Grafana dashboard as an admin under the `/grafana` subpath.
```

![Authentication overview](authentication-overview.png)

## Prerequisites

- You already have a working One Eye installation.
- If you want to use GitHub as a backend, register an OAuth app. For details, sees {{% xref "create-github-oauth.md" %}}.

## Steps

1. Install cert-manager.

    ```bash
    one-eye cert-manager install --update
    ```

1. Install an external ingress controller.

    ```bash
    one-eye ingress install --update
    ```

> 

Reviewer: Not yet released, but I've managed to make these obsolete in the current mainline, which we plan to release next week, before we release this doc.

```suggestion
```

1. Install Pomerium. The following command installs Pomerium and Dex, and configures Dex as well:

    ```bash
    one-eye pomerium install --update
    ```

    The installer will ask you for the following parameters:

>     1. **Root domain of Pomerium (localhost.banzaicloud.io)**: Enter the root domain for Pomerium. The default root domain is `localhost.banzaicloud.io`. If you don't use the default, note that:

Reviewer: ```suggestion
    1. **Root domain of Pomerium (localhost.banzaicloud.io)**: Enter the root domain for Pomerium. The default root domain is `localhost.banzaicloud.io`, which is good for experimenting locally, since it resolves to `127.0.0.1`. If you don't use the default, note that:
```

>         - You must configure your DNS to resolve `*.<root-domain>` addresses.

Reviewer: ```suggestion
        - You must configure your DNS to resolve `*.<root-domain>` addresses. (Or at least add `one-eye.<root-domain>`, `authentication.<root-domain>` and the `dex.<root-domain>` to your hosts file pointing to the external ingress' Loadbalancer address.)
```

        - GitHub authentication will require a custom clientID/secret pair too. For details, see {{% xref "/docs/one-eye/authentication/create-github-oauth.md" %}}.

    1. **Configure GitHub IdP? (Y/n)**: You can configure Dex to authenticate through a GitHub client, or you can use an automatically generated static login.

        - To use static login, enter N.

>             <!-- FIXME add how to retrieve generated login/password -->

Reviewer: This will be printed at the end of the command.
```suggestion
```

        - To use GitHub authentication, enter Y, then provide the GitHub client ID and GitHub Client Secret of the OAuth app you want to use. For details, see {{% xref "/docs/one-eye/authentication/create-github-oauth.md" %}}.

>             By default, everyone who successfully completes the authentication gets read and write access to One Eye. You can specify usergroups that get read (viewer) or read and write (editor) access in the **spec.authorization.groupRoleMapping** section of the [Observer custom resource]({{< relref "/docs/one-eye/crds/oneeye_types.md#observerspec" >}}).

Reviewer: ```suggestion
            By default, everyone who successfully completes the authentication needs to be explicitly allowed to pass Pomerium authorization, and additionally to have read and/or write access to One Eye. You can specify user groups that can pass through Pomerium and get read (viewer) or read and write (editor) access in the **spec.authorization.groupRoleMapping** section of the [Observer custom resource]({{< relref "/docs/one-eye/crds/oneeye_types.md#observerspec" >}}).
```

>             <!-- FIXME example -->

Reviewer: ```suggestion
apiVersion: one-eye.banzaicloud.io/v1alpha1
kind: Observer
metadata:
  name: one-eye
spec:
  authorization:
    groupRoleMapping:
      demo: viewer
      demo-editor: editor
```

> 

Reviewer: ```suggestion

        Individual users can also be allowlisted to pass Pomerium, but they will be authorized to access the One Eye dashboard based on the following conditions:
        - if `groupRoleMapping` is empty every user in `authorizedUsers` will have the `editor` role.
        - if `groupRoleMapping` has one or more items, then `authorizedUsers` will have roles that are defined by the mapping, so they will *not* have `editor` by default

```

    1. **Configure GitHub IdP? (Y/n)**: You can configure Dex to authenticate through a GitHub client, or you can use an automatically generated static login.

        - To use static login, enter N.
        - To use GitHub authentication, enter Y, then provide the GitHub client ID and GitHub Client Secret of the OAuth app you want to use. For details, see {{% xref "/docs/one-eye/authentication/create-github-oauth.md" %}}.

            By default, everyone who successfully completes the authentication needs to be explicitly allowed to pass Pomerium authorization, and additionally to have read and/or write access to One Eye. You can specify user groups that can pass through Pomerium and get read (viewer) or read and write (editor) access in the **spec.authorization.groupRoleMapping** section of the [Observer custom resource]({{< relref "/docs/one-eye/crds/oneeye_types.md#observerspec" >}}). For example:

            ```yaml
            apiVersion: one-eye.banzaicloud.io/v1alpha1
            kind: Observer
            metadata:
              name: one-eye
            spec:
              authorization:
                groupRoleMapping:

>                   demo: viewer
>                   demo-editor: editor

Reviewer: ```suggestion
                  example-user: viewer
                  example-admin: editor
```
